### ISSUE TITLE
Add `Task::park_timeout`

### ISSUE THREAD

**Issue by jonhoo** [rust-lang-nursery/futures-rs#198]
I've been rewriting my spmc-broadcast library [`bus`](https://github.com/jonhoo/bus) to expose a `Futures`-based interface rather than `recv/try_recv` (see the [`futures` branch](https://github.com/jonhoo/bus/tree/futures)), and am running into an issue. Due to the way the library works internally, there exists a condition under which the writer won't know that it needs to `unpark` the readers when it does a write. If this happens on the write that fills the buffer, the writer is not blocked on writing further, and the readers aren't aware that they need to start reading. 

In the non-futures version, this is "fixed" by parking the readers using `thread::park_timeout`. It's not elegant, but it avoids the otherwise costly solution necessary to fix the underlying issue. Unfortunately, with `Futures`, only `task::park` is available, leading to a deadlock in the library whenever the scenario above occurs.

Would adding `task::park_timeout` be possible in `Futures`? Are there any major drawbacks to doing so?

**Comment by alexcrichton** [comment]
Thanks for the report! 

This definitely seems like a nice-to-have and would also mirror `std::thread` quite nicely. Unfortunately I don't think it will be trivial to implement, but that doesn't mean it isn't possible! I think the most plausible route here is to track the earliest timeout requested in a `Task`, and then that could be read out of a spawned task. That way, if you're in an event loop like tokio-core, the event loop would poll a future, check the timeout, and if set would schedule a timer to get fired.

Unfortunately though implementing this would place a burden on downstream implementors of event loops as they'd now have to also implement timeouts, where that may not always be possible. 

Out of curiosity, what are you using for your event loop?

**Comment by jonhoo** [comment]
Ah, I see the issue. The `unpark` would need to be scheduled by the driving loop, whereas currently the loop doesn't even need to keep the `unpark` around. I've so far only really used `Futures` with `wait`, and so this has been a non-issue (and implementing `park_timeout` for `wait` should be straightforward).

It also seems useful for the event loop to give a future some kind of notion of time passing anyway; having this would enable us to have a timer future, similar to Go's `time.After`, which would be nice. Your proposed solution seems fine, though I wonder if we could make this a property of `Unpark` instead. That might make it easier for the event loops to automatically schedule the unparks.

With regards to event loops that cannot easily implement timeouts, I think we can document that `task::park_timeout` gives very weak guarantees about time (even weaker than `thread::park_timeout`): it only guarantees that the task will be woken up at *some* point in the future without `unpark` being called explicitly on the handle returned by `park_timeout`. The timeout is meant solely as a guideline to the outer loop that it is probably not useful to poll more often than the given timeout.

`bus`, linked to in the original post, is a library, and so does not implement an event loop instead. It constructs a broadcast channel, and implements `Future` (well, actually, `Stream`) for channel receivers. It leaves driving those `Future`s up to users of the library. The sender path attempts to `unpark` any parked receive tasks when it adds data, but unless I add some fairly expensive synchronization, there exists a race where the writer may not wake up some receivers.

**Comment by alexcrichton** [comment]
Ah ok yeah that makes sense. I'm a little hesitant to add this personally, but I think it'd be something we'd want to sort out for the 0.2 release.

I can see how it'd be *super* useful to have timeout futures and such in the core library as opposed to relying on timeouts being in external crates.

**Comment by alexcrichton** [comment]
Tagging as an 0.2-blocker as I think if we're gonna do this we should do it in 0.2. 

cc @carllerche

**Comment by carllerche** [comment]
Sounds interesting, but I need to let it marinate more.

Adding this would require every executor to be able to provide a timeout? What would the guarantees be? etc...

**Comment by alexcrichton** [comment]
Yeah I'm thinking that all tasks would have some sort of associated "next timeout" and an executor would have to be able to notify tasks when that timeout elapses. Whether that's *enough* to implement this I'm not so sure, but it's an idea at least perhaps.

It'd also put timeouts in futures core which'd be nice :)

**Comment by alexcrichton** [comment]
bah didn't mean to close.

**Comment by carllerche** [comment]
I don't think this is relevant anymore? Specifically, I think the route that will be explored is that `futures-timer` will provide a "global timer" similar to how Tokio's global reactor will work.

Can this be closed?

**Comment by alexcrichton** [comment]
I don't think that's true, this is still a route I think would be worth gaming out at some point.

**Comment by alexcrichton** [comment]
I'm going to close this as I think that we're probably not going to tackle this for 0.2 or the near future. While a nice-to-have it's also a pretty big ask for all executors to handle timeouts and also for futures themselves to juggle timeout requests. In the meantime I'm hoping that crates like `futures-timer` can be pretty slim in terms of dependencies and implementation to be a good general purpose timer for various needs.
